File: examples/cpp/dwarf_inspect.cpp
#include <LIEF/DWARF.hpp>
#include <LIEF/logging.hpp>
#include <LIEF/utils.hpp>

#include <cstdlib>
#include <iostream>

using namespace LIEF::logging;

int main(int argc, const char** argv) {
  if (!LIEF::is_extended()) {
    std::cerr << "This example requires the extended version of LIEF\n";
    return EXIT_FAILURE;
  }

  if (argc != 2) {
    std::cerr << "Usage: " << argv[0] << " <dwarf file>\n";
    return EXIT_FAILURE;
  }

  set_level(LEVEL::INFO);
  std::unique_ptr<LIEF::dwarf::DebugInfo> dbg = LIEF::dwarf::load(argv[1]);
  if (!dbg) {
    return EXIT_FAILURE;
  }
  for (std::unique_ptr<LIEF::dwarf::CompilationUnit> CU : dbg->compilation_units()) {
    log(LEVEL::INFO, "Producer: {}", CU->producer());
    for (std::unique_ptr<LIEF::dwarf::Function> func : CU->functions()) {
      log(LEVEL::INFO, "name={}, linkage={}, address={}",
          func->name(), func->linkage_name(), std::to_string(func->address().value_or(0)));
    }

    for (std::unique_ptr<LIEF::dwarf::Variable> var : CU->variables()) {
      log(LEVEL::INFO, "name={}, address={}", var->name(), std::to_string(var->address().value_or(0)));
    }

    for (std::unique_ptr<LIEF::dwarf::Type> ty : CU->types()) {
      log(LEVEL::INFO, "name={}, size={}", ty->name().value_or(""), std::to_string(ty->size().value_or(0)));
    }
  }

  dbg->find_function("_ZNSi4peekEv");
  dbg->find_function("std::basic_istream<char, std::char_traits<char> >::peek()");
  dbg->find_function(0x137a70);

  dbg->find_variable("_ZNSt12out_of_rangeC1EPKc");
  dbg->find_function("std::out_of_range::out_of_range(char const*)");
  dbg->find_function(0x2773a0);
  return EXIT_SUCCESS;
}

File: api/rust/examples/src/bin/dwarf_inspect.rs
/// This example shows how to inspect a DWARF debug info using LIEF's API

use std::process::{self, ExitCode};
use lief::dwarf::types::DwarfType;

fn main() -> ExitCode {
    if !lief::is_extended() {
        println!("This example requires the extended version of LIEF")
    }
    let mut args = std::env::args();
    if args.len() != 2 {
        println!("Usage: {} <binary>", args.next().unwrap());
        return ExitCode::FAILURE;
    }

    let path = std::env::args().last().unwrap();
    let mut file = std::fs::File::open(&path).expect("Can't open the file");

    let dbg = lief::dwarf::load(&path).unwrap_or_else(|| {
        process::exit(1);
    });

    for cu in dbg.compilation_units() {
        println!("Producer: {}", cu.producer());
        for func in cu.functions() {
            println!("name={}, linkage={}, address={}",
                func.name(), func.linkage_name(),
                func.address().unwrap_or(0)
            );
        }

        for var in cu.variables() {
            println!("name={}, address={}", var.name(), var.address().unwrap_or(0));
        }

        for ty in cu.types() {
            println!("name={}, size={}", ty.name().unwrap_or("".to_string()), ty.size().unwrap_or(0));
        }
    }

    dbg.function_by_name("_ZNSi4peekEv");
    dbg.function_by_name("std::basic_istream<char, std::char_traits<char> >::peek()");
    dbg.function_by_addr(0x137a70);

    dbg.variable_by_name("_ZNSt12out_of_rangeC1EPKc");
    dbg.variable_by_name("std::out_of_range::out_of_range(char const*)");
    dbg.variable_by_addr(0x137a70);

    println!("Can't process {}", path);
    ExitCode::FAILURE
}
